swupd/rename_test.go

```diff
@@ -0,0 +1,151 @@
+
+	sn := make(map[string]File)
+	for i := range files {
+		f := &files[i]
+		hi := HashFileInfo{Mode: f.m, Size: f.s}
+		contents := make([]byte, f.s)           // Load of NUL of correct length
+		copy(contents[0:len(f.k)], []byte(f.k)) // set the first n bytes
+		hv, err := GetHashForBytes(&hi, contents)
+		if err != nil {
+			t.Fatalf("GetHashForBytes for %v returned %v", f, err)
+		}
+		f.f.Hash = internHash(hv)
+		f.f.Info = sizer(f.s) // A FileInfo that returns our desired size
+		sn[f.n] = f.f
+	}
+	return sn
+}
+
+// filelist takes a list of wanted shortnames, looks them up in sn, and returns
+// a list of pointers to copies of the Files. Because they are copies they can
+// be altered without side effects on other tests
+func filelist(t *testing.T, sn map[string]File, wanted []string) []*File {
+	r := make([]*File, len(wanted))
+	for i := range wanted {
+		element, ok := sn[wanted[i]]
+		if !ok {
+			t.Fatalf("Shortname %v not found in array", wanted[i])
+		}
+		r[i] = &element
+	}
+	return r
+}
+
+// markdelete sets all the files in the slice to deleted
+func markdelete(a []*File) []*File {
+	for i := range a {
+		a[i].Status = StatusDeleted
+	}
+	return a
+}
+
+// checkLinked tests forward and backward pairing of DeltaPeers
+func checkLinked(t *testing.T, a, b *File, description string) bool {
+	// a is never nil, it is the entry in the "from" array
+	if a.DeltaPeer != b {
+		t.Errorf("%s Incorrect Linkage forward %v -> %v, got %v", description, a, b, a.DeltaPeer)
+		return false
+	}
+	// b is nil in the case that it is not supposed to have a DeltaPeer
+	if b != nil && b.DeltaPeer != a {
+		t.Errorf("%s Incorrect Linkage reverse %v -> %v, got %v", description, b, a, b.DeltaPeer)
+		return false
+	}
+	return true
+}
+
+func TestRename(t *testing.T) {
+	sn := generateTestArray(t)
+	tests := []struct {
+		description string
+		remove      []string // shortnames of File to remove
+		add         []string // shortnames of File to add
+		partner     []int    // Indexes of add that should match remove
+	}{
+		{description: "Rename of file, same contents",
+			remove: []string{"I1", "I2"}, add: []string{"I2"}, partner: []int{0, -1}},
+		{description: "Rename of short file, different contents, shouldn't link",
+			remove: []string{"S1"}, add: []string{"S2"}, partner: []int{-1}},
+		{description: "Rename of long file, different contents, should link",
+			remove: []string{"S1", "L1"}, add: []string{"L2"}, partner: []int{-1, 0}},
+		{description: "Rename of long files pairwise, should link",
+			remove: []string{"S1", "L1", "L3"}, add: []string{"L2", "L4"},
+			partner: []int{-1, 0, 1}},
+		{description: "Rename of long files pairwise reversed should link",
+			remove: []string{"S1", "L3", "L1"}, add: []string{"L2", "L4"},
+			partner: []int{-1, 1, 0}},
+		{description: "Rename files pairwise longer reversed should link",
+			remove: []string{"S1", "L3", "S2", "L1"}, add: []string{"L4", "L2"},
+			partner: []int{-1, 0, -1, 1}},
```

Comment:
Optional: it is easy to get confused with partner indices. I would consider making the expected result look more like "L3->L4 L1->L2" or ["L3->L4", "L1->L2"]. It is slightly more work in the test mechanics but I think pays off later.

Reply:
I agree it is easy to get confused with partner indicies, and yes, I did think about expressing the renames differently. I will try implementing it and see how it goes.